File: Gheetah/Templates/Java/Base_Desktop/src/test/java/{{ProjectName}}/stepdefinitions/DesktopSteps.java
package {{ProjectName}}.stepdefinitions;

import {{ProjectName}}.drivers.DriverFactory;
import io.appium.java_client.AppiumBy;
import io.appium.java_client.windows.WindowsDriver;
import io.cucumber.java.en.*;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import static org.assertj.core.api.Assertions.assertThat;
import java.time.Duration;

public class DesktopSteps {
    private WindowsDriver driver;
    private WebDriverWait wait;

    @Given("I launch desktop application at {string}")
    public void launchApp(String appPath) throws Exception {
        driver = DriverFactory.createDriver(appPath);
        wait = new WebDriverWait(driver, Duration.ofSeconds(15));
    }

    @When("I click on desktop element with {string} {string}")
    public void clickElement(String type, String val) {
        getEl(type, val).click();
    }

    @When("I type {string} into desktop element with {string} {string}")
    public void typeText(String text, String type, String val) {
        WebElement el = getEl(type, val);
        el.clear();
        el.sendKeys(text);
    }

    @When("I double click on desktop element with {string} {string}")
    public void doubleClick(String type, String val) {
        new Actions(driver).doubleClick(getEl(type, val)).perform();
    }

    @When("I maximize the desktop window")
    public void maximizeWindow() {
        driver.manage().window().maximize();
    }

    @Then("Desktop element with {string} {string} should exist")
    public void elementExists(String type, String val) {
        assertThat(getEl(type, val).isDisplayed()).isTrue();
    }

    @Then("Desktop element with {string} {string} text should be {string}")
    public void verifyText(String type, String val, String expected) {
        assertThat(getEl(type, val).getText()).isEqualTo(expected);
    }

    // Helper: Windows UI Automation Seçicileri
    private WebElement getEl(String type, String val) {
        org.openqa.selenium.By by = switch (type.toUpperCase()) {
            case "NAME" -> AppiumBy.name(val);
            case "ID" -> AppiumBy.accessibilityId(val);
            case "XPATH" -> AppiumBy.xpath(val);
            case "CLASS" -> AppiumBy.className(val);
            default -> AppiumBy.accessibilityId(val);
        };
        return wait.until(ExpectedConditions.visibilityOfElementLocated(by));
    }
}

File: Gheetah/Templates/C#/Base_Desktop/StepDefinitions/DesktopSteps.cs
using OpenQA.Selenium.Appium.Windows;
using OpenQA.Selenium.Interactions;
using Reqnroll;
using FluentAssertions;
using {{ProjectName}}.Hooks;
using System.Linq;

namespace {{ProjectName}}.StepDefinitions
{
    [Binding]
    public class DesktopSteps
    {
        private readonly WindowsDriver _driver;

        public DesktopSteps()
        {
            _driver = (WindowsDriver)Hooks.Hooks.Driver!;
        }

        [Given(@"I launch the desktop application at path ""(.*)""")]
        public void LaunchApp(string appPath)
        {
            if (Hooks.Hooks.Driver == null)
            {
                Hooks.Hooks.Driver = DriverFactory.CreateDriver(appPath);
            }
        }

        [When(@"I maximize the desktop window")]
        public void Maximize() => _driver.Manage().Window.Maximize();

        [When(@"I minimize the desktop window")]
        public void Minimize() => _driver.Manage().Window.Minimize();

        [When(@"I restore the desktop window")]
        public void Restore() => _driver.Manage().Window.Size = new System.Drawing.Size(1024, 768); // veya custom

        [When(@"I switch to window with Title ""(.*)""")]
        public void SwitchToWindow(string title)
        {
            var handle = _driver.WindowHandles.FirstOrDefault(h =>
            {
                _driver.SwitchTo().Window(h);
                return _driver.Title.Contains(title, StringComparison.OrdinalIgnoreCase);
            });

            handle.Should().NotBeNull($"Window with title containing '{title}' not found.");
            _driver.SwitchTo().Window(handle!);
        }

        [When(@"I close current window")]
        public void CloseWindow() => _driver.Close();

        // Click
        [When(@"I click desktop element with (Name|AccessibilityId|XPath|ClassName) ""(.*)""")]
        public void ClickDesktop(string type, string val) => GetElement(type, val).Click();

        [When(@"I right click desktop element with (Name|AccessibilityId|XPath) ""(.*)""")]
        public void RightClickDesktop(string type, string val)
        {
            var element = GetElement(type, val);
            new Actions(_driver).ContextClick(element).Perform();
        }

        // Input
        [When(@"I type ""(.*)"" into desktop element (Name|AccessibilityId|XPath) ""(.*)""")]
        public void TypeIntoDesktop(string text, string type, string val)
        {
            var el = GetElement(type, val);
            el.Clear();
            el.SendKeys(text);
        }

        [When(@"I clear desktop element (Name|AccessibilityId|XPath) ""(.*)""")]
        public void ClearDesktopElement(string type, string val) => GetElement(type, val).Clear();

        // Keyboard
        [When(@"I press (Enter|Tab|Escape|Space|F1|F2|F3|F4|F5|Delete) key")]
        public void PressKey(string key)
        {
            var keys = key switch
            {
                "Enter" => Keys.Enter,
                "Tab" => Keys.Tab,
                "Escape" => Keys.Escape,
                "Space" => Keys.Space,
                "Delete" => Keys.Delete,
                "F1" => Keys.F1,
                "F2" => Keys.F2,
                "F3" => Keys.F3,
                "F4" => Keys.F4,
                "F5" => Keys.F5,
                _ => throw new ArgumentException("Unsupported key")
            };
            new Actions(_driver).SendKeys(keys).Perform();
        }

        // Assertions
        [Then(@"Desktop element with (Name|AccessibilityId|XPath) ""(.*)"" should (exist|be visible|be enabled)")]
        public void AssertDesktopElement(string type, string val, string state)
        {
            var element = GetElement(type, val);
            switch (state)
            {
                case "exist":
                case "be visible":
                    element.Should().NotBeNull();
                    element.Displayed.Should().BeTrue();
                    break;
                case "be enabled":
                    element.Enabled.Should().BeTrue();
                    break;
            }
        }

        [Then(@"Desktop element with (Name|AccessibilityId) ""(.*)"" text should (be|contain) ""(.*)""")]
        public void AssertDesktopText(string type, string val, string opt, string expected)
        {
            var actual = GetElement(type, val).Text.Trim();
            if (opt == "be") actual.Should().Be(expected);
            else actual.Should().Contain(expected);
        }

        private WindowsElement GetElement(string type, string val)
        {
            return type switch
            {
                "Name" => _driver.FindElementByName(val),
                "AccessibilityId" => _driver.FindElementByAccessibilityId(val),
                "XPath" => _driver.FindElementByXPath(val),
                "ClassName" => _driver.FindElementByClassName(val),
                _ => throw new ArgumentException($"Invalid selector type: {type}")
            };
        }
    }
}